# Data Extraction from MIMIC.

### Contents:

* (A) Locating the required data
* (B) Extracting the data
* (C) Post-processing of data (involves combining with cohort information from 'cohort_selection_mimic.ipynb')

We use two mysql libraries:

* 'sql' is simple for scripting, with line/cell magic. 
* 'MySQLdb' is required for more complex operations (e.g. parameter binding with python lists).


In [1]:
import graphlab
import numpy as np
import pickle
from collections import OrderedDict

In [2]:
%load_ext sql
%sql mysql://root:mysql2016@localhost/MIMIC?unix_socket=/run/mysqld/mysqld.sock
%sql USE MIMIC

/home/rusty/anaconda3/lib/python2.7/site-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated since IPython 4.0. You should import from traitlets.config instead.
  "You should import from traitlets.config instead.", ShimWarning)
/home/rusty/anaconda3/lib/python2.7/site-packages/IPython/utils/traitlets.py:5: UserWarning: IPython.utils.traitlets has moved to a top-level traitlets package.
  warn("IPython.utils.traitlets has moved to a top-level traitlets package.")


This non-commercial license of GraphLab Create for academic use is assigned to cm1788@bristol.ac.uk and will expire on October 04, 2019.


[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1553264788.log


0 rows affected.


/home/rusty/anaconda3/lib/python2.7/site-packages/sqlalchemy/dialects/mysql/base.py:1569: Warning: '@@tx_isolation' is deprecated and will be removed in a future release. Please use '@@transaction_isolation' instead
  cursor.execute('SELECT @@tx_isolation')


[]

### (A) Locating the required data.

A number of different variables are required to assess the nurse led discharge (NLD) criteria. Here we locate the ITEMIDs corresponding to these variables, which will be used to extract patient data from chartevents.

For simplicity we focus on (dbsource=)'Metavision' only since data storage within 'Carevue' is less coherent. Because the study compares with a general intensive care unit (GICU) from the UK, we take only patients with (first_careunit=)'MICU' or 'SICU' since this patient subset approximately corresponds to the GICU population.

The required variables are:

#### To locate these variables we proceed as follows: 

We define these search terms:

And search for ITEMIDS with LABELS that match the search terms using the following query (replacing the search terms for each variable with those above):

In [3]:
%sql SELECT * FROM D_ITEMS WHERE (LABEL LIKE '%fraction%' OR LABEL LIKE '%fio2%') AND DBSOURCE='metavision'

6 rows affected.


[(12804, 223835, 'Inspired O2 Fraction', 'FiO2', 'metavision', 'chartevents', 'Respiratory', 'None', 'Numeric', None, 26L),
 (14083, 225628, 'CK-MB fraction (%)', 'CK-MB fraction (%)', 'metavision', 'chartevents', 'Labs', 'None', 'Numeric', None, None),
 (14515, 226754, 'FiO2ApacheIIValue', 'FiO2ApacheIIValue', 'metavision', 'chartevents', 'Scores - APACHE II', '%', 'Numeric', None, 26L),
 (14685, 227008, 'Ejection Fraction', 'Ejection Fraction', 'metavision', 'chartevents', 'Scores - APACHE IV (2)', '%', 'Numeric', None, None),
 (14686, 227009, 'FiO2_ApacheIV_old', 'FiO2_ApacheIV_old', 'metavision', 'chartevents', 'Scores - APACHE IV (2)', 'None', 'Numeric', None, None),
 (14687, 227010, 'FiO2_ApacheIV', 'FiO2_ApacheIV', 'metavision', 'chartevents', 'Scores - APACHE IV (2)', '%', 'Numeric', None, None)]

For each variable we visually inspect the search ouptut and select the ITEMS that appear relevant. 

*For example, for FiO2 we select 226754, 227009, 227010,223835 as candidates.*

We then measure the approximate frequnecy of the candidate variables on a random subset of the data (i.e. how many ICUSTAYS have at least one recording of each ITEMID)...

In [4]:
sub_limit = 100  ## subset size: number of ICUSTAYS randomly selected  

In [5]:
%sql SELECT D.ITEMID, D.LABEL, COUNT(DISTINCT(II.ICUSTAY_ID)) AS STAY_COUNT, (COUNT(DISTINCT(II.ICUSTAY_ID))/(:sub_limit)) AS FREQ \
FROM CHARTEVENTS C \
INNER JOIN ( \
    SELECT * FROM ( \
        SELECT * \
        FROM ICUSTAYS I \
        WHERE I.DBSOURCE='metavision' AND (I.FIRST_CAREUNIT='MICU' or I.FIRST_CAREUNIT='SICU') \
        ORDER BY RAND() \
        LIMIT :sub_limit) AS II_sub \
    ORDER BY II_sub.SUBJECT_ID, II_sub.HADM_ID, II_sub.ICUSTAY_ID \
            ) AS II \
ON (C.SUBJECT_ID=II.SUBJECT_ID \
    AND C.HADM_ID=II.HADM_ID \
    AND C.ICUSTAY_ID=II.ICUSTAY_ID) \
INNER JOIN D_ITEMS D \
ON C.ITEMID=D.ITEMID \
WHERE D.ITEMID IN (226754, 227009, 227010, 223835) \
GROUP BY D.ITEMID

1 rows affected.


[(223835, 'Inspired O2 Fraction', 47L, Decimal('0.4700'))]

..and manually inspect the first few instances of each ITEM to confirm it contains the required data:

In [6]:
%sql SELECT * FROM CHARTEVENTS WHERE ITEMID=223835 LIMIT 5 

5 rows affected.


[(335L, 34, 144319, 290505, 223835, datetime.datetime(2191, 2, 23, 7, 31), datetime.datetime(2191, 2, 23, 7, 35), 16924, '60', 60.0, None, 0, 0, None, None),
 (430L, 34, 144319, 290505, 223835, datetime.datetime(2191, 2, 23, 11, 0), datetime.datetime(2191, 2, 23, 11, 4), 14913, '60', 60.0, None, 0, 0, None, None),
 (706L, 36, 165660, 241249, 223835, datetime.datetime(2134, 5, 12, 7, 9), datetime.datetime(2134, 5, 12, 7, 9), 17525, '100', 100.0, None, 0, 0, None, None),
 (789L, 36, 165660, 241249, 223835, datetime.datetime(2134, 5, 12, 12, 0), datetime.datetime(2134, 5, 12, 13, 56), 17525, '100', 100.0, None, 0, 0, None, None),
 (877L, 36, 165660, 241249, 223835, datetime.datetime(2134, 5, 12, 16, 13), datetime.datetime(2134, 5, 12, 16, 27), 18428, '100', 100.0, None, 0, 0, None, None)]

*For example ITEMID 223835 (Inspired O2 Fraction) contains numeric fractional data, and is measured with a reasonably high frequency (~0.4 to 0.6). We connclude that this is a relevant ITEM.*	

#### By the above procedure we identify the following ITEMIDS corresponding to the required variables:

* Some ITEMIDS may not be in (frequent) use in the database - we select them anyway for completeness.
* Urine output may be located in either CHARTEVENTS or OUTPUTEVENTS (we check both).

**As a single list this is** (220739, 223900, 223901, 226755, 226756, 226757, 226758, 227011, 227012, 227013, 227014,228112, 220615, 226752, 227005, 223791, 227881, 227519, 227059, 220640, 227464, 227442, 226772, 226535, 220645, 226534, 226776, 224826, 226759, 227443,220227, 220277, 226860,226861,226862,226863,226865,228232, 225624, 227000, 227001, 223838, 224832, 224391, 227810,223837, 224829, 226754, 227009, 227010,223835, 220210, 224688, 224689, 224690, 226770,227039,227516,220224,220235,226062,226063,227036, 223761, 223762, 224027, 220045, 220228, 220339, 224699, 224700, 220050, 220059, 220179, 224167, 225309, 227243, 226850, 226852, 228151) ** for CHARTEVENTS.**

** And ** (226566, 226627, 226631) **for OUTPUTEVENTS (urine output only).**

-------------------------------------
## (B) Extracting the data

Having located the relevant ITEMIDs we now:

* Calculate their usage frequency from a larger random sample of ICUSTAYS
* Define a cuttoff frequency
* Extract and save all measurements of the remaining variables

In [7]:
sub_limit=1000  ## size of random sample

As before we calculate how many ICUSTAYS have at least one measurement of each ITEM, and convert this to an occurence frequnecy (FREQ). This time we calculate for all ITEMIDs simultaneoulsy, and sort the results by ascending occurence frequency. For CHARTEVENTS:

In [8]:
%sql SELECT D.ITEMID, D.LABEL, COUNT(DISTINCT(II.ICUSTAY_ID)) AS STAY_COUNT, (COUNT(DISTINCT(II.ICUSTAY_ID))/(:sub_limit)) AS FREQ \
FROM CHARTEVENTS C \
INNER JOIN ( \
    SELECT * FROM ( \
        SELECT * \
        FROM ICUSTAYS I \
        WHERE I.DBSOURCE='metavision' AND (I.FIRST_CAREUNIT='MICU' or I.FIRST_CAREUNIT='SICU') \
        ORDER BY RAND() \
        LIMIT :sub_limit) AS II_sub \
    ORDER BY II_sub.SUBJECT_ID, II_sub.HADM_ID, II_sub.ICUSTAY_ID \
            ) AS II \
ON (C.SUBJECT_ID=II.SUBJECT_ID \
    AND C.HADM_ID=II.HADM_ID \
    AND C.ICUSTAY_ID=II.ICUSTAY_ID) \
INNER JOIN D_ITEMS D \
ON C.ITEMID=D.ITEMID \
WHERE D.ITEMID IN (220739, 223900, 223901, 226755, 226756, 226757, 226758, 227011, 227012, 227013, 227014,228112, 220615, 226752, 227005, 223791, 227881, 227519, 227059, 220640, 227464, 227442, 226772, 226535, 220645, 226534, 226776, 224826, 226759, 227443,220227, 220277, 226860,226861,226862,226863,226865,228232, 225624, 227000, 227001, 223838, 224832, 224391, 227810,223837, 224829, 226754, 227009, 227010,223835, 220210, 224688, 224689, 224690, 226770,227039,227516,220224,220235,226062,226063,227036, 223761, 223762, 224027, 220045, 220228, 220339, 224699, 224700, 220050, 220059, 220179, 224167, 225309, 227243, 226850, 226852, 228151) \
GROUP BY D.ITEMID \
ORDER BY FREQ

43 rows affected.


[(227243, 'Manual Blood Pressure Systolic Right', 5L, Decimal('0.0050')),
 (220059, 'Pulmonary Artery Pressure systolic', 8L, Decimal('0.0080')),
 (224167, 'Manual Blood Pressure Systolic Left', 11L, Decimal('0.0110')),
 (228232, 'PAR-Oxygen saturation', 22L, Decimal('0.0220')),
 (223762, 'Temperature Celsius', 40L, Decimal('0.0400')),
 (227516, 'PO2 (Mixed Venous)', 46L, Decimal('0.0460')),
 (225309, 'ART BP Systolic', 54L, Decimal('0.0540')),
 (224829, 'Trach Tube Type', 55L, Decimal('0.0550')),
 (226062, 'Venous CO2 Pressure', 120L, Decimal('0.1200')),
 (226063, 'Venous O2 Pressure', 120L, Decimal('0.1200')),
 (227810, 'ETT Re-taped', 134L, Decimal('0.1340')),
 (226534, 'Sodium (whole blood)', 152L, Decimal('0.1520')),
 (224391, 'ETT Mark (location)', 159L, Decimal('0.1590')),
 (220227, 'Arterial O2 Saturation', 197L, Decimal('0.1970')),
 (227464, 'Potassium (whole blood)', 204L, Decimal('0.2040')),
 (224700, 'Total PEEP Level', 254L, Decimal('0.2540')),
 (220050, 'Arterial Blood Pressure systolic', 319L, Decimal('0.3190')),
 (223838, 'ETT Location', 334L, Decimal('0.3340')),
 (224832, 'ETT Type', 334L, Decimal('0.3340')),
 (223837, 'ETT Size (ID)', 335L, Decimal('0.3350')),
 (224688, 'Respiratory Rate (Set)', 335L, Decimal('0.3350')),
 (224690, 'Respiratory Rate (Total)', 365L, Decimal('0.3650')),
 (220339, 'PEEP set', 378L, Decimal('0.3780')),
 (224689, 'Respiratory Rate (spontaneous)', 378L, Decimal('0.3780')),
 (220224, 'Arterial O2 pressure', 460L, Decimal('0.4600')),
 (220235, 'Arterial CO2 Pressure', 460L, Decimal('0.4600')),
 (223835, 'Inspired O2 Fraction', 487L, Decimal('0.4870')),
 (223791, 'Pain Level', 894L, Decimal('0.8940')),
 (220228, 'Hemoglobin', 965L, Decimal('0.9650')),
 (227443, 'HCO3 (serum)', 967L, Decimal('0.9670')),
 (220615, 'Creatinine', 968L, Decimal('0.9680')),
 (220645, 'Sodium (serum)', 968L, Decimal('0.9680')),
 (225624, 'BUN', 968L, Decimal('0.9680')),
 (227442, 'Potassium (serum)', 968L, Decimal('0.9680')),
 (220179, 'Non Invasive Blood Pressure systolic', 979L, Decimal('0.9790')),
 (224027, 'Skin Temperature', 979L, Decimal('0.9790')),
 (223761, 'Temperature Fahrenheit', 985L, Decimal('0.9850')),
 (223900, 'GCS - Verbal Response', 985L, Decimal('0.9850')),
 (223901, 'GCS - Motor Response', 986L, Decimal('0.9860')),
 (220277, 'O2 saturation pulseoxymetry', 987L, Decimal('0.9870')),
 (220739, 'GCS - Eye Opening', 987L, Decimal('0.9870')),
 (220045, 'Heart Rate', 988L, Decimal('0.9880')),
 (220210, 'Respiratory Rate', 988L, Decimal('0.9880'))]

..and for OUTPUTEVENTS:

In [9]:
%sql SELECT D.ITEMID, D.LABEL, COUNT(DISTINCT(II.ICUSTAY_ID)) AS STAY_COUNT, (COUNT(DISTINCT(II.ICUSTAY_ID))/(:sub_limit)) AS FREQ \
FROM OUTPUTEVENTS C \
INNER JOIN ( \
    SELECT * FROM ( \
        SELECT * \
        FROM ICUSTAYS I \
        WHERE I.DBSOURCE='metavision' AND (I.FIRST_CAREUNIT='MICU' or I.FIRST_CAREUNIT='SICU') \
        ORDER BY RAND() \
        LIMIT :sub_limit) AS II_sub \
    ORDER BY II_sub.SUBJECT_ID, II_sub.HADM_ID, II_sub.ICUSTAY_ID \
            ) AS II \
ON (C.SUBJECT_ID=II.SUBJECT_ID \
    AND C.HADM_ID=II.HADM_ID \
    AND C.ICUSTAY_ID=II.ICUSTAY_ID) \
INNER JOIN D_ITEMS D \
ON C.ITEMID=D.ITEMID \
WHERE D.ITEMID IN (226566, 226627, 226631) \
GROUP BY D.ITEMID \
ORDER BY FREQ

2 rows affected.


[(226631, 'PACU Urine', 15L, Decimal('0.0150')),
 (226627, 'OR Urine', 139L, Decimal('0.1390'))]

**Note:** we find that the 'Urine Output' ITEMs that we identified are not in frequent use. The difficulty of identifying urine in MIMIC has been previoulsy noted by others. We accept that we canot use 'Urine Output' in our analysis and do not attempt further investigation.  

We now store the CHARTEVENTS query results into a dataframe, and remove those that are only recorded infrequently. This is done to simplify the subsequent analysis (and reduce the required number of checks on data integrity), but we aknowledge that it will result in a small amount of missing data. We define a minimum usage frequency of 0.01 (1%).

In [10]:
freqs=%sql SELECT D.ITEMID, D.LABEL, COUNT(DISTINCT(II.ICUSTAY_ID)) AS STAY_COUNT, (COUNT(DISTINCT(II.ICUSTAY_ID))/(:sub_limit)) AS FREQ \
FROM CHARTEVENTS C \
INNER JOIN ( \
    SELECT * FROM ( \
        SELECT * \
        FROM ICUSTAYS I \
        WHERE I.DBSOURCE='metavision' AND (I.FIRST_CAREUNIT='MICU' or I.FIRST_CAREUNIT='SICU') \
        ORDER BY RAND() \
        LIMIT :sub_limit) AS II_sub \
    ORDER BY II_sub.SUBJECT_ID, II_sub.HADM_ID, II_sub.ICUSTAY_ID \
            ) AS II \
ON (C.SUBJECT_ID=II.SUBJECT_ID \
    AND C.HADM_ID=II.HADM_ID \
    AND C.ICUSTAY_ID=II.ICUSTAY_ID) \
INNER JOIN D_ITEMS D \
ON C.ITEMID=D.ITEMID \
WHERE D.ITEMID IN (220739, 223900, 223901, 226755, 226756, 226757, 226758, 227011, 227012, 227013, 227014,228112, 220615, 226752, 227005, 223791, 227881, 227519, 227059, 220640, 227464, 227442, 226772, 226535, 220645, 226534, 226776, 224826, 226759, 227443,220227, 220277, 226860,226861,226862,226863,226865,228232, 225624, 227000, 227001, 223838, 224832, 224391, 227810,223837, 224829, 226754, 227009, 227010,223835, 220210, 224688, 224689, 224690, 226770,227039,227516,220224,220235,226062,226063,227036, 223761, 223762, 224027, 220045, 220228, 220339, 224699, 224700, 220050, 220059, 220179, 224167, 225309, 227243, 226850, 226852, 228151) \
GROUP BY D.ITEMID \
ORDER BY FREQ

freqs=graphlab.SFrame(freqs.DataFrame())
freqs.print_rows(num_rows=4)
ITEMIDS = freqs[freqs['FREQ']>=0.01]['ITEMID']  ## cutoff = 0.01
ITEMIDS_TOREMOVE = freqs[freqs['FREQ']<0.01]['ITEMID']  ## remove these from variable mapping

61 rows affected.
+--------+---------------------+------------+-------+
| ITEMID |        LABEL        | STAY_COUNT |  FREQ |
+--------+---------------------+------------+-------+
| 227000 |     BUN_ApacheIV    |     1      | 0.001 |
| 227001 |  BunScore_ApacheIV  |     1      | 0.001 |
| 227005 | Creatinine_ApacheIV |     1      | 0.001 |
| 227010 |    FiO2_ApacheIV    |     1      | 0.001 |
+--------+---------------------+------------+-------+
[61 rows x 4 columns]



In [11]:
print "There are %d infrequent ITEMS to remove from the variable mapping." %len(ITEMIDS_TOREMOVE)

There are 21 infrequent ITEMS to remove from the variable mapping.


In [12]:
variable_mapping = dict()

variable_mapping['fio2'] = [226754, 227009, 227010,223835]
variable_mapping['resp'] = [220210, 224688, 224689, 224690]
variable_mapping['po2'] = [226770,227039,227516,220224]  ## 227516 is venous - excluded in subsequent script.
variable_mapping['pco2'] = [220235,226062,226063,227036]  ## 226062,226063 are venous - excluded in subsequent script.

variable_mapping['temp'] = [223761, 223762, 224027] 
variable_mapping['hr'] = [220045]
variable_mapping['bp'] = [220050, 220059, 220179, 224167, 225309, 227243, 226850, 226852, 228151] ## 228151 is diastolic - excluded in subsequent script.
variable_mapping['k'] = [220640, 227464, 227442, 226772, 226535]
variable_mapping['na'] = [220645, 226534, 226776]
variable_mapping['hco3'] = [224826, 226759, 227443]
variable_mapping['spo2'] = [220227, 220277, 226860,226861,226862,226863,226865,228232]
variable_mapping['bun'] = [225624, 227000, 227001]
variable_mapping['airway'] = [223838, 224832, 224391, 227810,223837, 224829]
variable_mapping['gcs'] = [220739, 223900, 223901, 226755, 226756, 226757, 226758, 227011, 227012, 227013, 227014,228112]
variable_mapping['creatinine'] = [220615, 226752, 227005]
variable_mapping['pain'] = [223791, 227881]
variable_mapping['urine'] = [227519, 227059]
variable_mapping['haemoglobin'] = [220228]
variable_mapping['peep'] = [220339, 224699, 224700]
                        
for var in variable_mapping:
        variable_mapping[var] = [itd for itd in variable_mapping[var] if itd not in ITEMIDS_TOREMOVE]
        
print variable_mapping

{'urine': [], 'pain': [223791], 'temp': [223761, 223762, 224027], 'hr': [220045], 'fio2': [227009, 223835], 'resp': [220210, 224688, 224689, 224690], 'airway': [223838, 224832, 224391, 227810, 223837, 224829], 'po2': [226770, 227039, 227516, 220224], 'hco3': [224826, 227443], 'peep': [220339, 224699, 224700], 'gcs': [220739, 223900, 223901, 227011, 227012, 227014, 228112], 'pco2': [220235, 226062, 226063], 'na': [220645, 226534], 'bun': [225624], 'bp': [220050, 220179, 225309, 226850, 226852, 228151], 'creatinine': [220615], 'k': [220640, 227464, 227442, 226535], 'spo2': [220227, 220277, 226860, 226861, 226862, 226863, 226865, 228232], 'haemoglobin': [220228]}


#### Finally we extract the data from CHARTEVENTS for all the selected ITEMIDS:

In [13]:
import MySQLdb
import MySQLdb.cursors
from contextlib import closing
import shutil

def sql_connect():
    db='MIMIC'
    user='root'
    password='mysql2016'

    connection = MySQLdb.connect(host="localhost",
                     user=user, 
                     passwd=password, 
                     db=db,
                     unix_socket="/run/mysqld/mysqld.sock",
                     cursorclass = MySQLdb.cursors.SSCursor)  ## ensures correct behaviour for 'fetchone'
    return connection

In [14]:
## test the connection (should only print one row if 'fetchone' is working correctly):
conn = sql_connect()
with closing(conn.cursor()) as cur:
    cur.execute('SELECT * FROM D_ITEMS WHERE LABEL LIKE "%urine%" AND DBSOURCE="metavision"')
    row = cur.fetchone()
    print row
conn.close()

(12710, 220799, 'ZSpecific Gravity (urine)', 'ZSpecific gravity (urine)', 'metavision', 'chartevents', 'Labs', 'None', 'Numeric', None, None)


In [15]:
list_of_ids = ITEMIDS
format_strings = ','.join(['%s'] * len(list_of_ids))

In [16]:
partial_pull_query0 ="""SELECT C.SUBJECT_ID, C.HADM_ID, C.ICUSTAY_ID, C.ITEMID, C.CHARTTIME, C.VALUE, C.VALUENUM, C.VALUEUOM, II.INTIME, II.OUTTIME, II.LOS
FROM CHARTEVENTS C 
INNER JOIN ICUSTAYS II
ON (C.SUBJECT_ID=II.SUBJECT_ID 
    AND C.HADM_ID=II.HADM_ID 
    AND C.ICUSTAY_ID=II.ICUSTAY_ID)  
WHERE II.DBSOURCE='metavision' 
AND (II.FIRST_CAREUNIT='MICU' or II.FIRST_CAREUNIT='SICU') 
AND II.ICUSTAY_ID=%d"""

partial_pull_query1="""
   AND C.ITEMID in """

We now extract the data. This is a big query. We use 'fetchone' one to grab the results row by row, and store them in a dictionary (keys=column names, values=list of column values).

We use graphlab because is has out-of-memory storage for SFrames (no concern over size of result set). Still, we select the data patient by patient and build up the results frame incrementaly.

Note the SFrame.from_sql() method with MysqlDB connection was found to be more robust than sqlalchemy and pandas (using read_sql_query).

In [17]:
with closing(sql_connect()) as conn: 
    df = graphlab.SFrame.from_sql(conn, "SELECT DISTINCT(ICUSTAY_ID) FROM ICUSTAYS WHERE DBSOURCE='metavision' AND (FIRST_CAREUNIT='MICU' OR FIRST_CAREUNIT='SICU')", dbapi_module=MySQLdb, cursor_arraysize=1000)
    icustays = list(df['ICUSTAY_ID'])
    print "There are %d icustays in MICU and SICU under metavision." %len(icustays)

There are 14595 icustays in MICU and SICU under metavision.


We proceed in batches of 3000 patients and then clean up at the end (after ~5000 graphlab seem to start thrashing).

### Store ICUSTAY_IDs that produce errors and re-run for these! Sometimes the mySQL service dies on the laptop and takes some time ot restart. Hopefully it is just this!

#### See batch 1 : it seemed to be icustay that triggered the error. Are the subsequent errors for consecutive ICUstays while the service restarts?

#### If so we could avoid these errors by testing the connection and then waiting for the service to restart.....

In [ ]:
icustays = [i for i in icustays if i not in (249026, 254172)]

In [ ]:
icustays_with_sql_errors = [] 
## cannot find a totally robust solution - some icustays also crash 
## (it may be to do with the MySQL service on this computer, or my sql driver). 
## In either case, we store the erroneous ids and re-run the data extraction for these stays variable by variable.

for batch in [0,1,2,3,4]:
    
    print "Batch " + str(batch)
    first_icid_index = 0 + batch*3000
    last_icid_index = min(3000 + batch*3000, len(icustays))
    
    with closing(sql_connect()) as conn:
        pull_query = partial_pull_query0 %icustays[first_icid_index] + partial_pull_query1 + str(tuple(list_of_ids))
        results = graphlab.SFrame.from_sql(conn, pull_query, dbapi_module=MySQLdb, cursor_arraysize=1000)

    print "    Initialised batch results frame, proceeding with batch queries..."    
    for i,icid in enumerate(icustays[first_icid_index+1 : last_icid_index]):

        if i%500 == 0:
            print "%d %d" %(i, icid) 
        try:
            with closing(sql_connect()) as conn:
                pull_query = partial_pull_query0 %icid + partial_pull_query1 + str(tuple(list_of_ids))
                _df = graphlab.SFrame.from_sql(conn, pull_query, dbapi_module=MySQLdb, cursor_arraysize=1000)
        except:
            print "Sql error on icustay : %d" %icid
            icustays_with_sql_errors.append(icid)

        try:
            if len(_df)>0:
                results = results.append(graphlab.SFrame(_df))
        except:
            print "Could not join df to results on:  %d" %icid
    
    print "    Batch queries complete. Saving batch results..."
    results.save('_batch_%d_query_data' %batch)
    print "    Batch %d complete. Batch results contains %d rows." %(batch,len(results))

Batch 0
    Initialised batch results frame, proceeding with batch queries...
0 241249
500 280701
1000 218331
1500 278095
2000 216040
2500 239067
    Batch queries complete. Saving batch results...
    Batch 0 complete. Batch results contains 1821152 rows.
Batch 1
    Initialised batch results frame, proceeding with batch queries...
0 246007
500 277881
1000 293316
1500 244855
2000 237177
2500 292053
    Batch queries complete. Saving batch results...
    Batch 1 complete. Batch results contains 2549385 rows.
Batch 2
    Initialised batch results frame, proceeding with batch queries...
0 283328
500 292737
1000 243405
1500 298627
2000 236772
2500 235580
    Batch queries complete. Saving batch results...
    Batch 2 complete. Batch results contains 2342751 rows.
Batch 3
    Initialised batch results frame, proceeding with batch queries...
0 266340
500 274052
1000 261558
1500 248886
2000 276657
2500 234288
    Batch queries complete. Saving batch results...
    Batch 3 complete. Batch res

We now join the batch results and cleanup temporary files on the hardrive:

In [168]:
results = graphlab.SFrame('_batch_0_query_data')

for batch in [1,2,3,4]:
    _df = graphlab.SFrame('_batch_%d_query_data' %batch)
    results = results.append(_df)

results.save('mimic_all_data') ## rename: MIMIC_ALL_DATA

In [169]:
print len(results)

10853094


If joining the batched results was successful we can safely remove the batch files:

In [ ]:
for batch in [1,2,3,4]:
    shutil.rmtree('_batch_%d_query_data' %batch)

We now need to extract D.Label and D.Unitname from D_Items and join them to the results table:

In [ ]:
ditems_query = """SELECT ITEMID AS C.ITEMID, 
                         LABEL AS D.LABEL
                         UNITNAME AS D.UNITNAME
                         FROM D_ITEMS
                         WHERE ITEMID IN """

item_info = graphlab.SFrame.from_sql(conn, ditems_query + str(tuple(list_of_ids), dbapi_module=MySQLdb, cursor_arraysize=1000)

In [ ]:
results = results.join(item_info, on='C.ITEMID', how='inner')

In [ ]:
results.save('_test_all_query_data')

We now re-run the data extraction for the icustays that cause sql errors. If they fail again we will either invesitgate further and just exclude them from the data set. 

In [165]:
proceed = False
try:
    with closing(sql_connect()) as conn:
        pull_query = partial_pull_query0 %icustays_with_sql_errors[-1] + partial_pull_query1 + str(tuple(list_of_ids))
        results = graphlab.SFrame.from_sql(conn, pull_query, dbapi_module=MySQLdb, cursor_arraysize=1000)
        proceed = True
except:
    print "SQL error repeated for icustay %d." %icid

if proceed:
    for icid in reversed(icustays_with_sql_errors[0:-1]):
        try:
            with closing(sql_connect()) as conn:
                pull_query = partial_pull_query0 %icid + partial_pull_query1 + str(tuple(list_of_ids))
                _df = graphlab.SFrame.from_sql(conn, pull_query, dbapi_module=MySQLdb, cursor_arraysize=1000)
        except:
            print "SQL error repeated for icustay %d." %icid

        try:
            if len(_df)>0:
                results = results.append(_df)
        except:
            print "Could not join df to results on:  %d" %icid
            
    results.save("_batch_repeat_erroneous_stays")

SQL error repeated for icustay 249026.
SQL error repeated for icustay 254172.


Exception _mysql_exceptions.OperationalError: (2013, 'Lost connection to MySQL server during query') in <bound method SSCursor.__del__ of <MySQLdb.cursors.SSCursor object at 0x7f214e0f1fd0>> ignored


#### Why do some icustays always cause it to crash? (Since it is only two we assume these entries are corrupted and remove them from the analysis. They are only 8 and 6 days long.).

In [167]:
with closing(sql_connect()) as conn:
    pull_query = "SELECT LOS FROM ICUSTAYS WHERE ICUSTAY_ID=%d" %254172
    _df = graphlab.SFrame.from_sql(conn, pull_query, dbapi_module=MySQLdb, cursor_arraysize=1000)
    
print _df

+--------+
|  LOS   |
+--------+
| 5.9247 |
+--------+
[1 rows x 1 columns]



In [170]:
with closing(sql_connect()) as conn:
    pull_query = partial_pull_query0 %249026 + partial_pull_query1 + str(tuple(variable_mapping['fio2']))
    _df = graphlab.SFrame.from_sql(conn, pull_query, dbapi_module=MySQLdb, cursor_arraysize=1000)

Exception _mysql_exceptions.OperationalError: (2013, 'Lost connection to MySQL server during query') in <bound method SSCursor.__del__ of <MySQLdb.cursors.SSCursor object at 0x7f214c7ab1d0>> ignored


OperationalError: (2006, 'MySQL server has gone away')

#### There seems to be one patient near the end of batch 1 with a huge amount of data?

## Compare from here on:

In [175]:
table = graphlab.SFrame('mimic_all_data/')
old_table = graphlab.SFrame('mimic_all_data_CLEANED_RFD/')

IN_IDS = table['C.ICUSTAY_ID'].unique()
O_IN_IDS = old_table['C.ICUSTAY_ID'].unique()

In [176]:
remaining_ids = [iid for iid in O_IN_IDS if iid not in IN_IDS]

In [177]:
print len(IN_IDS)
print len(O_IN_IDS)

14430
11288


### Appears to have worked, but cannot confirm until see how many rows in the cleaned dataset.

----------------------------------------------------------------------------------------------
## (C) Post-processing of data.

Having extracted the main data we perform some post-processing to facilitate analysis. 

In [178]:
all_data = graphlab.SFrame('mimic_all_data') ## reload
print all_data.column_names()

number_of_stays = len(all_data['C.ICUSTAY_ID'].unique())
print "There are %d rows in the data. " %len(all_data)
print "There are %d intensive care stays." %number_of_stays

['C.CHARTTIME', 'C.HADM_ID', 'C.ICUSTAY_ID', 'C.ITEMID', 'C.SUBJECT_ID', 'C.VALUE', 'C.VALUENUM', 'C.VALUEUOM', 'D.LABEL', 'D.UNITNAME', 'II.INTIME', 'II.LOS', 'II.OUTTIME']
10841059
14430


*There are more than 10 million rows in the data (on disk this is just over 1.5gb saved). And there are 14430 unique ICUSTAYS, not all of these will be in the cohort).*

We add the following columns to the data:

* ['final_4hr'] : 1 if measurement is from final 4 hours of patients stay
* ['final_24hr'] : 1 if measurement is from final 24 hours of patients stay
* ['hrs_bd'] : float giving number of hours before discharge that measurement was taken (can filter on this column later)

In [179]:
all_data['hrs_bd'] = (all_data['II.OUTTIME'] - all_data['C.CHARTTIME'])/float(60**2)
all_data.save('mimic_all_data')  

In [180]:
HR = 4  ## number of hours before end of ICUSTAY 
all_data['final_%dhr' %HR] = all_data.apply(lambda row: 1 if (row['II.OUTTIME'] - row['C.CHARTTIME']).total_seconds()/(60.**2) <= HR else 0)
all_data.save('mimic_all_data')            

In [181]:
HR = 24  ## number of hours before end of ICUSTAY 
all_data['final_%dhr' %HR] = all_data.apply(lambda row: 1 if (row['II.OUTTIME'] - row['C.CHARTTIME']).total_seconds()/(60.**2) <= HR else 0)
all_data.save('mimic_all_data')            

#### Merging data extraction with cohort..

We now add columns for cohort identification, and filtering based on outcome. Main cohort consists of the first ICUSTAY of each hospital admission (should be >13000 stays).

From this cohort we are interested in ICUSTAYS that have good (1) and bad (0) outcome.
The are two types of bad outcome: (A) in-hospital death, (B) readmission to ICU (during same hospital admission).

The above abstraction is not fool-proof. For example:
* patients may die outside of hospital for related reasons
* readmission may be linked but occur on different stay
* in-hopistal death may be unrelated to ICU discharge
* etc

We add the following columns to the data frame: 

* ['cohort']: binary flag. 1 indicates that ICUSTAY is part of cohort (i.e. first or only stay of a hospital admission).
* ['outcome']: binary flag. 1 indicates good outcome (survival with no readmission). 0 indicates bad outcome (death or readmission).
* ['in_h_death']: binary flag. 1 for death. 0 for survival.
* ['in_icu_death']: binary flag. 1 for death. 0 for survival.
* ['readmit']: integer. Number of readmissions to icu during same hospital admission.

*Note: there are fewer HADMs than in cohort_selection_mimic.ipynb, since not all patients have chartevents data.*

In [182]:
f = open('first_stays.pkl', 'rb')
first_stays = pickle.load(f)
f.close()

f = open('moratlities.pkl', 'rb')
mortalities = pickle.load(f)
f.close()

f = open('stay_counts.pkl', 'rb')
stay_counts = pickle.load(f)
f.close()

f = open('icu_deaths.pkl', 'rb')
ic_deaths = pickle.load(f)
f.close()

In [183]:
print "Adding cohort column..."
all_data['cohort'] = all_data.apply(lambda row: 1 if first_stays[row['C.HADM_ID']]==row['C.ICUSTAY_ID'] else 0)

print "Adding in hospital column..."
all_data['in_h_death'] = all_data['C.HADM_ID'].apply(lambda hadmid: 1 if hadmid in mortalities else 0 )

print "Adding in icu death column..."
all_data['in_icu_death'] = all_data['C.ICUSTAY_ID'].apply(lambda icustay: 1 if icustay in ic_deaths else 0 )

print "Adding readmission column..."
all_data['readmit'] = all_data['C.HADM_ID'].apply(lambda hadmid: stay_counts[hadmid] - 1)

print "Adding outcome column..."
all_data['outcome'] = all_data['C.HADM_ID'].apply(lambda hadmid: 1 if (hadmid not in mortalities and stays_counts[hadmid]==1) else 0)

print 'finish adding new columns, saving data frame..'
all_data.save('mimic_all_data')

Adding cohort column...
Adding in hospital column...
Adding in icu death column...
Adding readmission column...
Adding outcome column...
finish adding new columns, saving data frame..


In [184]:
print "There are %d unique ICU stays, of which" %len(all_data['C.ICUSTAY_ID'].unique())
print "%d are in the cohort." %(sum(all_data['cohort']==1))

14430